Implement a better options menu

Polarite opened the pull request:
Currently doesnt have a menu for resource packs. 
Theres an issue with updating the clouds when you change settings, I dont think its an issue on options menu side.
Its missing some options that would be nice to have, open for suggestions on that.
Chat settings should have a proper textbox implementation for setting width...

Polarite committed:
WIP First draft of a... functioning settings menu and fixes to cloud fullbright.

Theres currently many issues but I taped it together and it kinda works(?)

 it currently has these problems:
* Doesnt include all options
* Cloud options when changed override "moving clouds" setting.
* Sliders sometimes dont set to exact numbers D:
* Slider system isnt exactly same as minecrafts, which I am not sure how to make work... It doesnt feel as smooth as minecrafts sliders...
* The Chat width and height text boxes cant be changed, the boxes needs a special input block similar to... you guessed it!
* The tooltips dont follow the cursor, I couldnt find out a way to implement it that isnt a taped together ugly tech.
* Sliders are missing the grayed out textures, currently only making them slightly transparent, at least til I can bother to fix the "knob" texture being an actual knob :')
* Settings tabs should probably be reorganized in a better manner, current organization feels a little all over the place.
* Translation strings needs to be added, I gave up at some point :')
* I dont have headphones with me... Does the master volume slider work??
*  Graying out of buttons doesnt work after rebasing.
* Hi, please dont be mean :)

Update SliderElement.kt

fixed the updated version of minosoft (graying out buttons is broken)

Polarite committed:
Update AbstractButtonElement.kt

fixes to make it actually compile, damn it.

Polarite committed:
Simplify settings menus using Menu base class and fix cloud fullbright

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/elements/input/button/AbstractButtonElement.kt:
```diff
@@ -113,8 +114,11 @@ abstract class AbstractButtonElement(
-        background.render(offset, consumer, options)
-        textElement.render(offset + Vec2f(HorizontalAlignments.CENTER.getOffset(size.x, textSize.x), VerticalAlignments.CENTER.getOffset(size.y, textSize.y)), consumer, options)
+        // Apply reduced opacity when disabled
+        val renderOptions = if (disabled) GUIVertexOptions(alpha = 0.4f) else options
```
render options are inherited. You need to use `options.copy`

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/elements/input/button/AbstractButtonElement.kt:
```diff
@@ -180,15 +184,15 @@ abstract class AbstractButtonElement(
 
     private companion object {
-        val CLICK_SOUND = minecraft("ui.button.click")
+        val CLICK_SOUND = "minecraft:ui.button.click".toResourceLocation()
```
`        val CLICK_SOUND = minecraft("ui.button.click")`

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/elements/input/slider/SliderElement.kt:
```diff
@@ -0,0 +1,156 @@
+        val size = size
+        val sliderHeight = 20.0f
+
+        // Render background (button texture)
```
no need for those self explaining comments

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/elements/input/slider/SliderElement.kt:
```diff
@@ -0,0 +1,156 @@
+
+        // Calculate slider position (0.0 to 1.0)
+        val normalizedValue = (value - min) / (max - min)
+        val sliderWidth = 8.0f
```
const/option?

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/elements/input/slider/SliderElement.kt:
```diff
@@ -0,0 +1,156 @@
+        if (isDragging) {
+            updateValueFromPosition(position.x)
+        }
+        return true
```
the return value means that the event is consumed. Is it when not dragging?

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/elements/input/slider/SliderElement.kt:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    private fun updateValueFromPosition(x: Float) {
+        val sliderWidth = 8.0f
```
constant

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/gui/screen/menu/options/ChatSettingsMenu.kt:
```diff
@@ -0,0 +1,87 @@
+            chatProfile.hidden = !chatProfile.hidden
+            hiddenButton.textElement.text = formatEnabled("menu.options.chat.hidden", chatProfile.hidden)
+        }
+        this += hiddenButton
```
why not just `this += ButtonElement(...`

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/gui/screen/menu/options/GUISettingsMenu.kt:
```diff
@@ -0,0 +1,130 @@
+        updateDisabledStates()
+    }
+
+    private fun updateDisabledStates() {
```
The profile provides delegates. You can easily observe it. Justs create a profile synced button (or so), that observes the profile and updates the button when that "event" is triggered. No need for polling.

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/gui/screen/menu/options/OptionsMenu.kt:
```diff
@@ -0,0 +1,83 @@
+            guiRenderer.gui.push(ChatSettingsMenu)
+        }
+        this += ButtonElement(guiRenderer, "menu.options.language".i18n()) {
+            JavaFXUtil.runLater { Eros.setVisibility(true) }
```
there is something missing.

Bixilon reviewed (changes requested):
Just the initial code review, have not looked deeper inside nor tested it.

Polarite committed:
Added Language Settings, done most of the requested fixes.

Slider bars still are wonky, I will rewrite that.
Controls are still not working, I am unsure how can I make something that looks good with current UI and menus.
Do we... even support resource packs lmao?

Polarite commented:
Slider bars still are wonky, I will rewrite that. 
Controls are still not working, I am unsure how can I make something that looks good with current UI and menus.
Do we... even support resource packs...
Please check if I have fixed them like you want, I kinda got confused at some points. Language settings is dirty but its as best as I could manage to make it...

Polarite committed:
Fixes to cloud settings, Added controls settings, fixed slider element and mouse capturing for it.

Polarite committed:
Update copyright year

my bad for using outdated file haha :')

Polarite committed:
Make duplicate keybinds in controls change to red.

Polarite committed:
Fix bindingmanager getting confused when combination keys are used.

Polarite committed:
fix combination key holding in keyactionfilter.

Polarite committed:
Fixes to combination key displays and overriding keys if a combination can be fullfilled

This excludes cases if a combination for "E+T+Y" exists and you press "E" to open the inventory, you can press Y+T+E on those cases so neither chat or inventory opens and you can use the combination for that... I think even vanilla minecraft bothered with their controls menu this much I swear...

Polarite committed:
Implemented a textbox element and a search system for controls just in case.

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/gui/screen/menu/options/ControlsSettingsMenu.kt:
```diff
@@ -0,0 +1,823 @@
+        private val SCROLLBAR_TRACK_COLOR = RGBAColor(40, 40, 40, 180)
+        private val SCROLLBAR_THUMB_COLOR = RGBAColor(120, 120, 120, 220)
+        
+        val CATEGORIZED_KEYBINDINGS: Map<String, Map<ResourceLocation, KeyBinding>> = linkedMapOf(
```
why are all keybindings duplicated here?

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/input/MouseCapturing.kt:
```diff
@@ -0,0 +1,53 @@
+import de.bixilon.minosoft.gui.rendering.gui.input.mouse.MouseButtons
+
+/**
+ * Interface for GUI elements that can capture mouse input.
```
I really don't want such comments, everything should be self explaining .

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/input/MouseCapturing.kt:
```diff
@@ -0,0 +1,53 @@
+ * This file exists purely for sliders in menus to be able to capture mouse outside their bounds. 
+ * Could have more uses in future...
+ */
+interface MouseCapturing {
```
Why just relativeX? Why not relativeY? Why relative at all and not absolute?

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/input/key/manager/binding/BindingsManager.kt:
```diff
@@ -103,10 +103,72 @@ class BindingsManager(
     }
 
     fun onKey(code: KeyCodes, pressed: Boolean, handler: InputHandler?, time: ValueTimeMark) {
+        // Find all satisfied bindings that use this key and track their key counts
```
Not sure if thats good. The current behavior is depending on the order of registration. Shouldn't all of them just be triggered? But that breaks (e.g. Q and CTRL + Q). The whole keysystem feels somehow not great anymore.

This function needs refactoring, feels really long and clumsy now.

Polarite replied:
the problem with the current system was if you assign Shift+Q as a key for walking forward character drops item in inventory while achieving shift+q key as well, and if the combination keys are used for movement, and stopping pressing the keys isnt detected

Polarite replied:
I believe controlling in general needs a rewrite before properly supporting combination keystrokes...

Bixilon commented on src/main/java/de/bixilon/minosoft/gui/rendering/gui/gui/screen/menu/options/ControlsSettingsMenu.kt:
```diff
@@ -731,6 +780,18 @@ class ControlsSettingsMenu(guiRenderer: GUIRenderer) : Screen(guiRenderer), Abst
+             * Creates a unique signature for a key binding based on its key codes.
+             * Used for duplicate detection.
+             */
+            fun getBindingSignature(binding: KeyBinding): String {
```
Why is this a string? Why not just the set (you can use EnumSets from kutil)

Bixilon committed:
language: add integrated language as fallback

Bixilon committed:
option menu: button delegate

Bixilon reviewed (commented):
So (finally making a review):

All the possible languages are in code, that feels wrong. Can be move that to a json file in the assets? 371dc9aae1fb61520cb331a91ad2ad39e61d0a57


I don't really care about the old UI system anymore, so I am kind of blindly accepting it. 

From UI perspective:

 - feels weird that you can smoothly slide the biome radius like it was a decimal value.
 - scrolling in controls makes me change the hotbar slot


From code:
 - there is a lot of duplicated code or unused variables
 - please remove unnecessary comments

This style is not good:
```
    private val textFilteringButton: ButtonElement

...
        textFilteringButton = ButtonElement(guiRenderer, formatEnabled("menu.options.chat.text_filtering", chatProfile.textFiltering)) {
            chatProfile.textFiltering = !chatProfile.textFiltering
            textFilteringButton.textElement.text = formatEnabled("menu.options.chat.text_filtering", chatProfile.textFiltering)
        }
        this += textFilteringButton

```

This whole part can be squashed into
```
this += BooleanDelegateButton(guiRenderer, "menu.options.chat.text_filtering".i18n(), delegate = chatProfile::textFiltering)
```
(and it handles translation, getting, setting, observing and formatting)

I added those changes, see 0e7c250c5029895d4566bc59eec7810b143fcd7b


Translation are messed:
```
    protected fun translate(key: String): String {
        return IntegratedLanguage.LANGUAGE.forceTranslate(key.i18n().translationKey).message
    }
```

The text formatting does that automatically, no need for anything. just use the translation key as text (in a base component).


I don't really like that the in game ui is responsible for most settings, why not in eros?


This is a really big first PR, it might make sense to split this. 


Is this AI coded? I can not accept pure AI contributions.
